Fix Translation for alert banner not displaying

andybroomfield opened the pull request:
Fix #318 

Alters the alert banner block to check for a translation, and add a translation test.

Originally authoured by [adinancenci](https://www.drupal.org/u/adinancenci).
Original patch in comment 3.
  https://www.drupal.org/project/localgov/issues/3427615#comment-15487385
(Patch originallty placed the logic in the entity, moved to block to fix issue with visibility checking).

andybroomfield committed:
Fix Translation for alert banner not displaying

Originally authoured by [adinancenci](https://www.drupal.org/u/adinancenci).
Applies patch in comment 3.
  https://www.drupal.org/project/localgov/issues/3427615#comment-15487385

andybroomfield committed:
First pass at writing translation test

andybroomfield committed:
Second pass at translation test with support for paths

andybroomfield committed:
cs: fix up translation test

Wrap up translation test.

ekes commented on src/Plugin/Block/AlertBannerBlock.php:
```diff
@@ -192,8 +204,15 @@ protected function getCurrentAlertBanners() {
+
+      // Get alert banner translation if it exists.
+      if ($alert_banner->hasTranslation($langcode)) {
+        $alert_banner = $alert_banner->getTranslation($langcode);
```
Shouldn't this be EntityRepository::getTranslationFromContext()

and I assume this has knock on with the fact that this is called from the constructor (and probably shouldn't be).

andybroomfield replied:
Yeah I was looking into that, I think we might need to get #321 merged first to check this (remove call from constructer).

That said, the alert banner loaded is the correct translation until the condition fields display paths are set to something different for the translation, so I assume there is already some translation context aware loading going on.

I added a test case for the scenarios I was seeing the break on and the above method does work. Is there a particular reason to favour `EntityRepository::getTranslationFromContext()` as opposed to `$alert_banner->getTranslation($langcode)` method?

ekes replied:
> I added a test case for the scenarios I was seeing the break on and the above method does work. Is there a particular reason to favour EntityRepository::getTranslationFromContext() as opposed to $alert_banner->getTranslation($langcode) method?

Because the former is intended for getting translations to edit them, the later is intended for getting translations to display them in context should they exist. The later does the work of checking if there is a translation for you, checks cachability, and it uses the language fallback system (as configured for the particular site) to decide which language to use.

andybroomfield committed:
Replace getTranslation with entityRespositry::getTranslationFromContext

Use the entitiy repositry service to get translation in correct language
with fallback.

andybroomfield commented:
This is ready for re-review as I'd like to get this fixed and close the ticket on drupal.org.

ekes reviewed (approved):
Looks good :+1:

andybroomfield committed:
Move the language selection logic to the block

This is to allow the entitiy to remain pure in terms of logic, and the get
translation logic in the block. If the alert has a translation use that in place
of the alert banner entitiy. This also allows correct checking of visibility

Also allows the title to be translated.
